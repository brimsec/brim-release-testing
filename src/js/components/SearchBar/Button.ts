import {darken} from "polished"
import styled from "styled-components"

import {cssVar} from "../../lib/cssVar"

const Button = styled.button`
  background: none;
  padding: 0;
  border: none;
  position: relative;
  height: 28px;
  width: 28px;
  display: flex;
  align-items: center;
  justify-content: center;
  background: rgba(0, 0, 0, 0);
  border-radius: 50%;

  svg {
    height: 16px;
    width: 16px;
    fill: var(--slate);
  }

  &:not(:disabled) {
    &:hover {
      background: rgba(0, 0, 0, 0.05);
    }

    &:active {
      background: rgba(0, 0, 0, 0.08);
      svg {
        fill: ${darken(0.1, cssVar("--slate"))};
      }
    }
  }

  &:disabled svg {
    opacity: 0.24;
  }
`

export default Button
