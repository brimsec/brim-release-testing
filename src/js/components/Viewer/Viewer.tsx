import React, {useEffect, useRef, useState} from "react"

import {RowRenderer, ScrollPosition, ViewerDimens} from "../../types"
import {reactElementProps} from "../../test/integration"
import Chunk from "./Chunk"
import Chunker from "./Chunker"
import Header from "./Header"
import Log from "../../models/Log"
import ScrollHooks from "../../lib/ScrollHooks"
import * as Styler from "./Styler"
import TableColumns from "../../models/TableColumns"
import lib from "../../lib"
import useConst from "../hooks/useConst"
import {useDispatch} from "react-redux"
import History from "../../state/History"

type Props = {
  chunker: Chunker
  dimens: ViewerDimens
  tableColumns: TableColumns
  renderRow: RowRenderer
  logs: Log[]
  onLastChunk?: Function
  renderEnd: () => any
  scrollPos: ScrollPosition
  innerRef: any
  timeZone: string
  timeFormat: string
}

export default function Viewer(props: Props) {
  const dispatch = useDispatch()
  const [scrollLeft, setScrollLeft] = useState(0)
  const [chunks, setChunks] = useState(props.chunker.visibleChunks(0))
  const ref = useRef<HTMLDivElement>()
  function onScrollStart() {
    lib.doc.id("tooltip-root").style.display = "none"
  }

  function onScrollStop() {
    const view = ref.current
    lib.doc.id("tooltip-root").style.display = "block"
    if (view) dispatch(History.update({x: view.scrollLeft, y: view.scrollTop}))
  }

  const scrollHooks = useConst(null, () =>
    ScrollHooks.create(onScrollStart, onScrollStop)
  )

  function onScroll() {
    scrollHooks && scrollHooks()
    const view = ref.current
    if (view) {
      updateChunks(view.scrollTop)
      setScrollLeft(view.scrollLeft)
    }
  }

  function updateChunks(scrollTop) {
    const next = props.chunker.visibleChunks(scrollTop)
    if (!Chunker.isEqual(next, chunks)) setChunks(next)
  }

  useEffect(() => {
    const view = ref.current
    if (!view) return
    updateChunks(view.scrollTop)
    if (props.chunker.lastChunk() == chunks[chunks.length - 1]) {
      props.onLastChunk && props.onLastChunk()
    }
  })

  useEffect(() => {
    const view = ref.current
    if (!view) return
    if (props.scrollPos) {
      view.scrollTo(props.scrollPos.x, props.scrollPos.y)
    }
  }, [props.scrollPos])

  return (
    <div className="viewer" ref={props.innerRef} tabIndex={0}>
      <Header
        columns={props.tableColumns}
        dimens={props.dimens}
        scrollLeft={scrollLeft}
        {...reactElementProps("viewer_header")}
      />
      <div
        className="view"
        onScroll={onScroll}
        style={{
          width: props.dimens.viewWidth,
          height: props.dimens.viewHeight
        }}
        ref={ref}
      >
        <List {...{...props, chunks}} />
      </div>
    </div>
  )
}

type ListProps = {
  chunks: number[]
} & Props

const List = React.memo<ListProps>(function List(props) {
  return (
    <div
      className="list"
      style={Styler.list(props.dimens)}
      {...reactElementProps("viewer_results")}
    >
      {props.chunks.map((chunk) => (
        <Chunk
          columns={props.tableColumns}
          logs={props.logs}
          rows={props.chunker.rows(chunk)}
          key={chunk}
          rowRenderer={props.renderRow}
          dimens={props.dimens}
        />
      ))}
      {props.renderEnd()}
    </div>
  )
})
