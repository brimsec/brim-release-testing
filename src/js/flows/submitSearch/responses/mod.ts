import fsExtra from "fs-extra"
import path from "path"

export function response(name: string) {
  const string = fsExtra.readFileSync(path.join(__dirname, name), "utf-8")
  return string.split("\n\n\n").map((json) => JSON.parse(json))
}
