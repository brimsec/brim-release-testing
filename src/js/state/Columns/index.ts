import actions from "./actions"
import reducer from "./reducer"
import selectors from "./selectors"
import touch from "./touch"

export default {
  ...actions,
  ...selectors,
  touch,
  reducer
}
