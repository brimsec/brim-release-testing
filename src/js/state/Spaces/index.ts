import actionsFor from "./actionsFor"
import actions from "./actions"
import reducer from "./reducer"
import selectors from "./selectors"

export default {
  ...actions,
  ...selectors,
  actionsFor,
  reducer
}
