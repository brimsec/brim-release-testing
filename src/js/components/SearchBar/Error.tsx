import {useSelector} from "react-redux"
import React from "react"
import styled from "styled-components"

import SearchBar from "../../state/SearchBar"
import Warning from "../icons/warning-sm.svg"

const Wrap = styled.div`
  display: flex;
  align-items: center;
  color: white;
  background: var(--green);
  font-size: 11px;
  padding: 6px 12px;
  margin: 0 12px;
  box-shadow: 0 0.5px 2px 1px rgba(0, 0, 0, 0.15);
  border-radius: 0 0 3px 3px;

  svg {
    fill: white;
    height: 16px;
    width: 16px;
    margin-right: 12px;
  }
`

export default function Error() {
  const error = useSelector(SearchBar.getSearchBarError)
  if (!error) return null
  return (
    <Wrap>
      <div className="warning-icon">
        <Warning />
      </div>
      <span>{error}</span>
    </Wrap>
  )
}
