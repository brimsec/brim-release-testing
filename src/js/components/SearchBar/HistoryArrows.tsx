import {useDispatch, useSelector} from "react-redux"
import React from "react"
import styled from "styled-components"

import Back from "../icons/back-arrow.svg"
import Button from "./Button"
import Forward from "../icons/forward-arrow.svg"
import SearchBar from "../../state/SearchBar"
import Tab from "../../state/Tab"

const Wrap = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`

export default function HistoryArrows() {
  const dispatch = useDispatch()
  const canGoBack = useSelector(Tab.canGoBack)
  const canGoForward = useSelector(Tab.canGoForward)
  const back = () => dispatch(SearchBar.goBack())
  const forward = () => dispatch(SearchBar.goForward())

  return (
    <Wrap>
      <Button disabled={!canGoBack} onClick={back}>
        <Back />
      </Button>
      <Button disabled={!canGoForward} onClick={forward}>
        <Forward />
      </Button>
    </Wrap>
  )
}
