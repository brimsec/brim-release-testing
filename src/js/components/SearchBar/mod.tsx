import React from "react"
import styled from "styled-components"

import Pins from "./Pins"
import Error from "./Error"
import HistoryArrows from "./HistoryArrows"
import Input from "./Input"
import InputBackdrop from "./InputBackdrop"
import MenuAction from "./MenuAction"
import PinAction from "./PinAction"
import Spinner from "./Spinner"
import SubmitButton from "./SubmitButton"
import TargetSelect from "./TargetSelect"

const Wrap = styled.div`
  width: 100%;
  display: flex;
  align-items: flex-start;
  -webkit-app-region: no-drag;
  margin-bottom: 12px;

  & > * {
    margin-right: 8px;
    &:last-child {
      margin-right: 0;
    }
  }
`

const Group = styled.div<{flex: number}>`
  display: flex;
  flex: ${(p) => p.flex || "initial"};
  flex-direction: column;
`

const InputActions = styled.div`
  display: flex;
  margin-right: 8px;
`

export function SearchBar() {
  return (
    <Wrap className="search-bar">
      <HistoryArrows />
      <Group flex={1}>
        <InputBackdrop>
          <TargetSelect />
          <Input />
          <Spinner />
          <InputActions>
            <PinAction />
            <MenuAction />
          </InputActions>
          <SubmitButton />
        </InputBackdrop>
        <Error />
        <Pins />
      </Group>
    </Wrap>
  )
}
