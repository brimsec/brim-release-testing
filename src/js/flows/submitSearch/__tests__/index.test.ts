import {createZealotMock} from "zealot"

import {response} from "../responses/mod"
import {submitSearch} from "../mod"
import Columns from "../../../state/Columns"
import Current from "../../../state/Current"
import Handlers from "../../../state/Handlers"
import SearchBar from "../../../state/SearchBar"
import Spaces from "../../../state/Spaces"
import Viewer from "../../../state/Viewer"
import fixtures from "../../../test/fixtures"
import initTestStore from "../../../test/initTestStore"

const indexResp = response("index_search.txt")
const space = fixtures("space1")
const warningResp = response("search_warning.txt")

let store, zealot, dispatch, select
beforeEach(() => {
  zealot = createZealotMock()
  store = initTestStore(zealot)
  dispatch = store.dispatch
  select = (s: any) => s(store.getState())
  zealot.stubStream("archive.search", indexResp)
  store.dispatchAll([
    Current.setConnectionId("1"),
    Spaces.setDetail("1", space),
    Current.setSpaceId(space.id),
    SearchBar.changeSearchBarInput(":ip=192.168.0.1"),
    SearchBar.setTarget("index")
  ])
})
const submit = (...args) => dispatch(submitSearch(...args))

describe("table search", () => {
  test("zealot sends one request", async () => {
    await submit()

    const calls = zealot.calls("archive.search")
    expect(calls.length).toBe(1)
    expect(calls[0].args).toEqual(
      expect.objectContaining({patterns: [":ip=192.168.0.1"], spaceId: "1"})
    )
  })

  test("the table gets populated", async () => {
    await submit()
    expect(select(Viewer.getViewerRecords).length).toBe(4)
  })

  test("the table gets cleared", async () => {
    dispatch(
      Viewer.setRecords(undefined, [
        [{name: "clear", type: "string", value: "me"}]
      ])
    )
    submit()
    expect(select(Viewer.getViewerRecords)).toEqual([])
  })

  test("the table status updates", async () => {
    const promise = submit()
    expect(select(Viewer.getStatus)).toBe("FETCHING")
    expect(select(Viewer.getEndStatus)).toBe("FETCHING")
    await promise
    expect(select(Viewer.getStatus)).toBe("SUCCESS")
    expect(select(Viewer.getEndStatus)).toBe("COMPLETE")
  })

  test("registers a table request then cleans it up", async () => {
    const promise = submit()
    expect(select(Handlers.get)["Table"]).toEqual(
      expect.objectContaining({type: "SEARCH"})
    )
    await promise
    expect(select(Handlers.get)["Table"]).toBe(undefined)
  })

  test("aborts previous table request", async () => {
    const abort = jest.fn()
    dispatch(Handlers.register("Table", {type: "SEARCH", abort}))
    await submit()
    expect(abort).toHaveBeenCalledTimes(1)
  })

  test("sets the viewer columns", async () => {
    await submit()
    expect(select(Viewer.getColumns)).toMatchSnapshot()
  })

  test("sets the Columns columns", async () => {
    await submit()
    expect(select(Columns.getColumns)).toMatchSnapshot()
  })

  test("search warnings", async () => {
    zealot.stubStream("archive.search", warningResp)
    await submit()
    expect(select(SearchBar.getSearchBarError)).toBe(
      "Cut field boo not present in input"
    )
  })
})
