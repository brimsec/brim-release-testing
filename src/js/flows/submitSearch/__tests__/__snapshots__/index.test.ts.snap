// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table search sets the Columns columns 1`] = `
Object {
  "f85b7481593632c9884034d965640304": Object {
    "_log:zfile": Object {
      "isVisible": true,
    },
    "count:count": Object {
      "isVisible": true,
    },
    "key:addr": Object {
      "isVisible": true,
    },
  },
}
`;

exports[`table search sets the viewer columns 1`] = `
Object {
  "f85b7481593632c9884034d965640304": Array [
    Object {
      "name": "key",
      "type": "addr",
    },
    Object {
      "name": "count",
      "type": "count",
    },
    Object {
      "name": "_log",
      "type": "zfile",
    },
  ],
}
`;
